### ISSUE TITLE
DynamicMapping annotation should be applicable to any object field.

### ISSUE THREAD

**Issue by TEXHIK** [spring-projects/spring-data-elasticsearch#1767]
As of the [documentation](https://www.elastic.co/guide/en/elasticsearch/reference/current/dynamic.html#dynamic-inner-objects), it is possible to make different values for dynamicMapping setting in document and it's inner objects.
However, it is not possible to enable dynamicMapping for `@Field(type = FieldType.Object)`, because it is not an entity.
There should be a way to apply this setting on any property, not only for documents, as there many cases, when Object field is not an entity - Map, for example.

**Comment by sothawo** [comment]
You can do this by adding the `@DynamicMapping` annotation to the property. See https://github.com/spring-projects/spring-data-elasticsearch/blob/4.1.7/src/test/java/org/springframework/data/elasticsearch/core/index/MappingBuilderUnitTests.java#L896-L910 for an example

**Comment by TEXHIK** [comment]
I tried, but it is completely ignored, property just mapped as { "type": "object"} without any other settings. There is a check if the property is an entity, and if not - dynamic mapping not applied, and o check for annotation is done in that case.
P.S. My property is a map, not a class.

**Comment by sothawo** [comment]
Can you provide the entity definition for your case so that this can be reproduced?

**Comment by TEXHIK** [comment]
Here is my Field in code:
```
@Field(type = FieldType.Object)
@DynamicMapping(DynamicMappingValue.True)
private Map<String, Object> body;
```
This is field in elastic:
```
"body": {
    "type": "object"
},
```

And this is a check in MappingBuilder:220
```
if (!isCompletionProperty && property.isEntity() && hasRelevantAnnotation(property)) {

	if (fieldAnnotation == null) {
		return;
	}

	if (isNestedOrObjectProperty) {
		Iterator<? extends TypeInformation<?>> iterator = property.getPersistentEntityTypes().iterator();
		ElasticsearchPersistentEntity<?> persistentEntity = iterator.hasNext()
				? elasticsearchConverter.getMappingContext().getPersistentEntity(iterator.next())
				: null;

		mapEntity(builder, persistentEntity, false, property.getFieldName(), true, fieldAnnotation.type(),
				fieldAnnotation, property.findAnnotation(DynamicMapping.class));
		return;
	}
}
```
So as you can see, nothing Object-related applied at all, if the property isn't an entity. But it isn't.

**Comment by sothawo** [comment]
fixed in master an backported to 4.2.x, 4.1.x and 4.0.x
